File: src/content/docs/topics/cs/leetcode-150/sliding-window/076-minimum-window-substring.py
def min_window(s: str, t: str) -> str:
    pass  # TODO: implement

def _run_tests():
    assert min_window('ADOBECODEBANC', 'ABC') == 'BANC'
    assert min_window('a', 'a') == 'a'
    assert min_window('a', 'aa') == ''
    assert min_window('', 'a') == ''
    assert min_window('abc', '') == ''
    assert min_window('aa', 'aa') == 'aa'
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    min_window('a' * 5000 + 'b' * 5000, 'ab')
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf 10000-char string: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/sliding-window/076-minimum-window-substring.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func minWindow(_ s: String, _ t: String) -> String {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:5253d46c8dc7e7e6537b126c7c84d1534f9ea4284ebcd0d20fe9ad1b4153913a
    expectEqual(Solution().minWindow("ADOBECODEBANC", "ABC"), "BANC", "canonical-window")
    expectEqual(Solution().minWindow("a", "a"), "a", "exact-single-character")
    expectEqual(Solution().minWindow("a", "aa"), "", "insufficient-frequency")
    expectEqual(Solution().minWindow("aa", "aa"), "aa", "whole-string-window")
    expectEqual(Solution().minWindow("a", "b"), "", "target-not-present")
    // EXCLUDED_VECTOR empty-target: ["abc",""] | The target must contain at least one character.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
